extensions/ql-vscode/src/data-extensions-editor/generate-flow-model.ts

```diff
@@ -0,0 +1,146 @@
+import { CancellationToken } from "vscode";
+import { DatabaseItem } from "../local-databases";
+import { join } from "path";
+import { QueryRunner } from "../queryRunner";
+import { CodeQLCliServer } from "../cli";
+import { extLogger, TeeLogger } from "../common";
+import { definitions } from "./yaml";
+import { ProgressCallback } from "../progress";
+import { getOnDiskWorkspaceFolders } from "../helpers";
+import { ModeledMethod, ModeledMethodType } from "./modeled-method";
+
+class FlowModelGenerator {
+  constructor(
+    private readonly cli: CodeQLCliServer,
+    private readonly queryRunner: QueryRunner,
+    private readonly queryStorageDir: string,
+    private readonly qlDir: string,
+    private readonly databaseItem: DatabaseItem,
+    private readonly progress: ProgressCallback,
+    private readonly token: CancellationToken,
+  ) {}
+
+  async getAddsTo(
+    type: Exclude<ModeledMethodType, "none">,
+    queryName: string,
+    queryStep: number,
+  ): Promise<Array<[string, ModeledMethod]> | undefined> {
+    const definition = definitions[type];
+
+    const query = join(
+      this.qlDir,
+      this.databaseItem.language,
+      "ql/src/utils/modelgenerator",
+      queryName,
+    );
+
+    const queryRun = this.queryRunner.createQueryRun(
+      this.databaseItem.databaseUri.fsPath,
+      { queryPath: query, quickEvalPosition: undefined },
+      false,
+      getOnDiskWorkspaceFolders(),
+      undefined,
+      this.queryStorageDir,
+      undefined,
+      undefined,
+    );
+
+    const queryResult = await queryRun.evaluate(
+      ({ step, message }) =>
+        this.progress({
+          message: `Generating ${type} model: ${message}`,
+          step: queryStep * 1000 + step,
+          maxStep: 4000,
+        }),
+      this.token,
+      new TeeLogger(this.queryRunner.logger, queryRun.outputDir.logPath),
+    );
+
+    const bqrsPath = queryResult.outputDir.bqrsPath;
+
+    const bqrsInfo = await this.cli.bqrsInfo(bqrsPath);
+    if (bqrsInfo["result-sets"].length !== 1) {
+      void extLogger.log(
+        `Expected exactly one result set, got ${bqrsInfo["result-sets"].length}`,
+      );
+      return undefined;
+    }
+
+    const resultSet = bqrsInfo["result-sets"][0];
+
+    const decodedResults = await this.cli.bqrsDecode(bqrsPath, resultSet.name);
+
+    const results = decodedResults.tuples;
+
+    return results
+      .map((result): [string, ModeledMethod] | undefined => {
```

Comment:
The `[string, ModeledMethod]` is being used again and again. Is it time to name it and re-use it?